### PR TITLE
Java bindings for JSON reader support

### PR DESCRIPTION
**Pull request by wbo4958**
This PR is trying to add JSON reader support for java bindings.

### CODE REVIEW

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java [hunk 1]
```diff
@@ -21,21 +21,48 @@
...
   public static final class Builder  extends ColumnFilterOptions.Builder<JSONOptions.Builder> {
+    private boolean dayFirst = false;
+    private boolean lines = true;
+
+    public Builder withDayFirst(boolean dayFirst) {
+      this.dayFirst = dayFirst;
+      return this;
+    }
+
+    public Builder withLines(boolean lines) {
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/JSONOptions.java:58]: Need some javadocs on what these options do.

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 2]
```diff
@@ -797,6 +801,31 @@ public static Table readCSV(Schema schema, CSVOptions opts, HostMemoryBuffer buf
...
+   * @return the file parsed as a table on the GPU.
+   */
+  public static Table readJSON(Schema schema, File path) {
+    return readJSON(schema, JSONOptions.DEFAULT, path);
+  }
+
+  /**
+   * Read a JSON file.
+   * @param schema the schema of the file.  You may use Schema.INFERRED to infer the schema.
+   * @param opts various CSV parsing options.
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/Table.java:817]: not CSV

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 3]
```diff
@@ -797,6 +801,31 @@ public static Table readCSV(Schema schema, CSVOptions opts, HostMemoryBuffer buf
...
+   * @return the file parsed as a table on the GPU.
+   */
+  public static Table readJSON(Schema schema, JSONOptions opts, File path) {
+    return new Table(
+        readJSON(schema.getTypeIds(), schema.getTypeScales(),
+            path.getAbsolutePath(),
+            0, 0,
+            opts.isDayFirst(), opts.isLines()));
+  }
+
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/Table.java:828]: Are there APIs to read from a buffer? That is what we are going to need for Spark

File: java/src/test/java/ai/rapids/cudf/TableTest.java [hunk 4]
```diff
@@ -292,6 +293,26 @@ void testGetNumberOfColumns() {
...
+        .column(DType.INT32, "B")
+        .build();
+    JSONOptions opts = JSONOptions.builder()
+        .withLines(true)
+        .build();
+    try (Table table = Table.readJSON(schema, opts, TEST_SIMPLE_JSON_FILE)) {
+      HostColumnVector hc = table.getColumn(0).copyToHost();
+      HostColumnVector hcage = table.getColumn(1).copyToHost();
+      StringBuilder builder = new StringBuilder();
+      for (int i = 0; i < hc.getRowCount(); i++) {
```
Comment (revans2) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:309]: Need some actual asserts here.

File: java/src/test/java/ai/rapids/cudf/TableTest.java [hunk 5]
```diff
@@ -292,6 +293,26 @@ void testGetNumberOfColumns() {
     }
   }
 
+  @Test
+  void testReadJSON() {
+    Schema schema = Schema.builder()
+        .column(DType.STRING, "A")
+        .column(DType.INT32, "B")
```
Comment (revans2) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:300]: How do "A" and "B" line up with "name" and "age" in the file? This feels really confusing because Spark goes off of the actual key "name" and "age". I would expect "A" and "B" to be null after reading.

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java (hunk 1 above)
Comment (wbo4958) [reply java/src/main/java/ai/rapids/cudf/JSONOptions.java:75]: Done.

File: java/src/main/java/ai/rapids/cudf/Table.java (hunk 3 above)
Comment (wbo4958) [reply java/src/main/java/ai/rapids/cudf/Table.java:850]: Added the API for reading buffer

File: java/src/main/java/ai/rapids/cudf/Table.java (hunk 2 above)
Comment (wbo4958) [reply java/src/main/java/ai/rapids/cudf/Table.java]: Sry, the previous commit was a draft. Now it's ready for  review

File: java/src/test/java/ai/rapids/cudf/TableTest.java (hunk 4 above)
Comment (wbo4958) [reply java/src/test/java/ai/rapids/cudf/TableTest.java]: Done.

File: java/src/test/java/ai/rapids/cudf/TableTest.java (hunk 5 above)
Comment (wbo4958) [reply java/src/test/java/ai/rapids/cudf/TableTest.java]: Done.

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java [hunk 6]
```diff
@@ -18,24 +18,78 @@
...
+     * @param dayFirst true: DD/MM, false, MM/DD
+     * @return
+     */
+    public Builder withDayFirst(boolean dayFirst) {
+      this.dayFirst = dayFirst;
+      return this;
+    }
+
+    /**
+     * Whether to the file as a json object per line
```
Comment (firestarman) [inline java/src/main/java/ai/rapids/cudf/JSONOptions.java:71]:
```suggestion
     * Whether to read the file as a json object per line
```
?

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java [hunk 7]
```diff
@@ -18,24 +18,78 @@
...
+    @Override
+    public Builder includeColumn(Collection<String> names) {
+      throw new IllegalArgumentException("JSON reader didn't support column prune");
+    }
+
     public JSONOptions build() {
       return new JSONOptions(this);
     }
   }
+
```
Comment (firestarman) [inline java/src/main/java/ai/rapids/cudf/JSONOptions.java:94]: useless change

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 8]
```diff
@@ -797,6 +802,97 @@ public static Table readCSV(Schema schema, CSVOptions opts, HostMemoryBuffer buf
...
+   * Read JSON formatted data.
+   * @param schema the schema of the data. You may use Schema.INFERRED to infer the schema.
+   * @param opts various JSON parsing options.
+   * @param buffer raw UTF8 formatted bytes.
+   * @param offset the starting offset into buffer.
+   * @param len the number of bytes to parse.
+   * @return the data parsed as a table on the GPU.
+   */
+  public static Table readJSON(Schema schema, JSONOptions opts, byte[] buffer, long offset,
+                              long len) {
```
Comment (firestarman) [inline java/src/main/java/ai/rapids/cudf/Table.java:861]:
```suggestion
                               long len) {
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 9]
```diff
@@ -797,6 +802,97 @@ public static Table readCSV(Schema schema, CSVOptions opts, HostMemoryBuffer buf
...
+    if (len <= 0) {
+      len = buffer.length - offset;
+    }
+    assert len > 0;
+    assert len <= buffer.length - offset;
+    assert offset >= 0 && offset < buffer.length;
+    try (HostMemoryBuffer newBuf = HostMemoryBuffer.allocate(len)) {
+      newBuf.setBytes(0, buffer, offset, len);
```
Comment (firestarman) [inline java/src/main/java/ai/rapids/cudf/Table.java:869]: Do we really need to copy the data before reading to GPU ?

**Review by revans2 — APPROVED** [review]
This looks okay to me. I am a bit concerned about how this is going to interact with corner cases in Spark.  But it is a good step forward.

File: java/src/test/java/ai/rapids/cudf/TableTest.java (hunk 5 above)
Comment (revans2) [reply java/src/test/java/ai/rapids/cudf/TableTest.java]: Still a little confused.  Did the "A" and "B" work at all?

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java [hunk 10]
```diff
@@ -18,24 +18,78 @@
...
+    return dayFirst;
+  }
+
+  public boolean isLines() {
+    return lines;
+  }
+
+  @Override
+  String[] getIncludeColumnNames() {
+    throw new IllegalArgumentException("JSON reader didn't support column prune");
```
Comment (gerashegalov) [inline java/src/main/java/ai/rapids/cudf/JSONOptions.java:49]: here and other places: what is an illegal argument here? should it be UnsupportedOperationException?

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 11]
```diff
@@ -797,6 +802,97 @@ public static Table readCSV(Schema schema, CSVOptions opts, HostMemoryBuffer buf
...
+   * @param len the number of bytes to parse.
+   * @return the data parsed as a table on the GPU.
+   */
+  public static Table readJSON(Schema schema, JSONOptions opts, HostMemoryBuffer buffer,
+                              long offset, long len) {
+    if (len <= 0) {
+      len = buffer.length - offset;
+    }
+    assert len > 0;
+    assert len <= buffer.getLength() - offset;
```
Comment (gerashegalov) [inline java/src/main/java/ai/rapids/cudf/Table.java:889]: nit: inconsistent usage of field `length` via a getter whereas the code above uses it directly

File: java/src/test/java/ai/rapids/cudf/TableTest.java (hunk 5 above)
Comment (wbo4958) [reply java/src/test/java/ai/rapids/cudf/TableTest.java]: No, specifying "A" "B" will not work. The first WIP commit can work is because column names were not passed to lib cudf.

Sorry for causing the confusion.

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java (hunk 6 above)
Comment (wbo4958) [reply java/src/main/java/ai/rapids/cudf/JSONOptions.java:71]: Done. Thx

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java (hunk 7 above)
Comment (wbo4958) [reply java/src/main/java/ai/rapids/cudf/JSONOptions.java:94]: Done.

File: java/src/main/java/ai/rapids/cudf/Table.java (hunk 8 above)
Comment (wbo4958) [reply java/src/main/java/ai/rapids/cudf/Table.java:861]: Done.

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java (hunk 10 above)
Comment (wbo4958) [reply java/src/main/java/ai/rapids/cudf/JSONOptions.java:49]: Done. Thx

File: java/src/main/java/ai/rapids/cudf/Table.java (hunk 11 above)
Comment (wbo4958) [reply java/src/main/java/ai/rapids/cudf/Table.java:889]: Done. Thx

**Comment by wbo4958** [comment]
rerun tests

**Review by revans2 — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java [hunk 12]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- *  Copyright (c) 2019, NVIDIA CORPORATION.
+ *  Copyright (c) 2019-2021, NVIDIA CORPORATION.
```
Comment (gerashegalov) [inline java/src/main/java/ai/rapids/cudf/JSONOptions.java:3]: the PR crossed into 2022, update all the headers accordingly

**Review by revans2 — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Comment by wbo4958** [comment]
@gpucibot merge
